staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -550,10 +479,29 @@ func (w *watchCache) waitUntilFreshAndListItems(ctx context.Context, resourceVer
 		}
 		return w.store.List(), w.resourceVersion, "", nil
 	}()
-
+	result, err = filterPrefix(key, result)
```

Comment:
Adding it here means that both filtering and sorting will happen under lock, which we don't want.

Reply:
I suggest leaving the `WaitUntilFreshAndList` as is just factor out the filter function as you did.

Reply:
Nice catch. Forgot about deferred lock.